lib/SILOptimizer/Transforms/SILCodeMotion.cpp

```diff
@@ -292,6 +293,50 @@ void BBEnumTagDataflowState::handlePredSwitchEnum(SwitchEnumInst *S) {
                    "the switch_enum.");
 }
 
+void BBEnumTagDataflowState::handleStringCmp(SILBasicBlock *Pred) {
+  // Find all apply instructions
+  for (auto begin = Pred->begin(); begin != Pred->end(); ++begin) {
+    if (auto *AI = dyn_cast<ApplyInst>(begin)) {
+      if (auto *FN = dyn_cast<FunctionRefInst>(AI->getCalleeOrigin())) {
+        // Only keep going if this is an apply instruction
+        if (FN->getReferencedFunctionOrNull()->getName() != "$sSS2eeoiySbSS_SStFZ") continue;
+      } else continue;
+      
+      StringRef FirstArg; // Keep track of one of the arguments
+      
+      for (auto& Arg : AI->getArgumentOperands()) {
```

Comment:
1. Please create a helper function somewhere to check if a `SILValue` is a call to construct a constant string and, if so, to return the constructed string.
2. Please make that call properly look for calls to a specific function (or a function with the right `@_semantics` rather than assuming that all calls which take `StringLiteralInsts` as their first argument fit your conditions.
3. You can then just call that helper function on the two arguments to the comparison function instead of doing this loop.

Reply:
I was just about to mention the same thing as @rjmccall. You can use this helper function: https://github.com/apple/swift/blob/91669cec8a7d34fa603c836b64bfd679abb8f1ad/lib/SILOptimizer/Mandatory/OSLogOptimization.cpp#L111 to check if this is a call to a string initializer that takes a string literal. This function can be factored out and put in a common place. I think `swift/lib/SILOptimizer/Utils/ConstExpr.h/cpp` can have this utility function.

Reply:
@ravikandhadai and @rjmccall let me know what you think of `StringLiteralInfo`.